substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+					let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+						"staker should exist in VoterList, as per the contract \
+                            with staking.",
+					);
+
+					let stake_imbalance = stake_imbalance_of(prev_stake, voter_weight.into());
+
+					log!(
+						debug,
+						"on_stake_update: {:?} with {:?}. impacting nominations {:?}",
+						who,
+						stake_imbalance,
+						nominations,
+					);
+
+					// updates vote weight of nominated targets accordingly. Note: this will
+					// update the score of up to `T::MaxNominations` validators.
+					for target in nominations.into_iter() {
+						Self::update_target_score(&target, stake_imbalance);
+					}
+				},
+				StakerStatus::Validator => {
+					// validator is both a target and a voter.
+					let stake_imbalance = stake_imbalance_of(prev_stake, voter_weight.into());
+
+					Self::update_target_score(who, stake_imbalance);
+
+					let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+						"the staker should exit in VoterList, as per the \
+                            contract with staking.",
+					);
+				},
+				StakerStatus::Idle => (), // nothing to see here.
+			}
+		}
+	}
+
+	/// Fired when someone sets their intention to validate.
+	///
+	/// A validator is also considered a voter with self-vote and should also be added to
+	/// [`Config::VoterList`].
+	//
+	/// Note: it is assumed that `who`'s ledger staking state is updated *before* calling this
+	/// method.
+	fn on_validator_add(who: &T::AccountId, self_stake: Option<Stake<BalanceOf<T>>>) {
+		let self_stake = self_stake.unwrap_or_default().active;
+
+		if !T::TargetList::contains(who) {
+			T::TargetList::on_insert(who.clone(), self_stake)
+				.expect("staker does not exist in the list as per check above; qed.");
+		} else {
+			// if the target already exists in the list, it means that the target has been idle
+			// and/or dangling.
+			debug_assert!(
+				T::Staking::status(who) == Ok(StakerStatus::Idle) ||
+					T::Staking::status(who).is_err()
+			);
+
+			let self_stake = Self::to_vote_extended(self_stake);
+			Self::update_target_score(who, StakeImbalance::Positive(self_stake));
+		}
+
+		log!(debug, "on_validator_add: {:?}. role: {:?}", who, T::Staking::status(who),);
+
+		// a validator is also a nominator.
+		Self::on_nominator_add(who, vec![])
+	}
+
+	/// Fired when a validator becomes idle (i.e. chilling).
+	///
+	/// While chilled, the target node remains in the target list.
+	///
+	/// While idling, the target node is not removed from the target list but its score is
+	/// updated.
+	fn on_validator_idle(who: &T::AccountId) {
+		let self_stake = Self::weight_of(Self::active_vote_of(who));
+		Self::update_target_score(who, StakeImbalance::Negative(self_stake.into()));
+
+		// validator is a nominator too.
+		Self::on_nominator_idle(who, vec![]);
+
+		log!(debug, "on_validator_idle: {:?}, decreased self-stake {}", who, self_stake);
```

Comment:
This is a good mental check: all fns related to validators must also call into some internal function related to the self-nomination of the validator.